qiime2/core/type/signature.py

```diff
@@ -318,6 +318,155 @@ def _assert_valid_views(self, inputs, parameters, outputs):
                     " Pipelines do not support function annotations (found one"
                     " for parameter: %r)." % name)
 
+    def coerce_given_parameters(self, provenance, **user_input):
```

Comment:
It would be nice if these layers did not have to consider provenance, that feels like something that should stay (be trapped) in `action.py` since provenance is quite mutable whereas things in `core.type` are quite functional.

In the case of parameters, you can pretty much do that without any change here, but the inputs and outputs seem trickier. So probably not worth addressing right now.

I really wish something like this was straight-forward:

```python
map(invocation.inputs, lambda name, spec, value: value._view(...)
```
where this "layer" basically just resolves the difference between lists/dicts/singular elements and transforms between them as needed. Or perhaps there's some dictionary of generic "Argument" objects which you can apply transformations to and it internally magics away the distinction.